File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/CounterBetCategorizer.java
package cz.fb.manaus.reactor.categorizer;


import cz.fb.manaus.core.category.BetCoverage;
import cz.fb.manaus.core.category.categorizer.SettledBetCategorizer;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.SettledBet;
import cz.fb.manaus.core.model.Side;
import org.springframework.stereotype.Component;

import java.util.Map;
import java.util.Set;

@Component
public class CounterBetCategorizer implements SettledBetCategorizer {

    public static final String PREFIX = "counter_";

    @Override
    public boolean isSimulationSupported() {
        return false;
    }

    @Override
    public Set<String> getCategories(SettledBet settledBet, BetCoverage coverage) {
        var marketId = settledBet.getBetAction().getMarket().getId();
        var selectionId = settledBet.getSelectionId();
        var side = settledBet.getPrice().getSide();
        var counterSide = side.getOpposite();
        var bets = coverage.getBets(marketId, selectionId, counterSide);
        var avgCounter = bets.stream()
                .map(SettledBet::getPrice)
                .mapToDouble(Price::getPrice)
                .average();
        if (avgCounter.isPresent()) {
            var counterPrice = avgCounter.getAsDouble();
            var price = settledBet.getPrice().getPrice();
            var prices = Map.of(side, price, counterSide, counterPrice);
            if (Price.priceEq(counterPrice, price)) {
                return Set.of(PREFIX + "zero");
            } else if (prices.get(Side.BACK) > prices.get(Side.LAY)) {
                return Set.of(PREFIX + "profit");
            } else if (prices.get(Side.BACK) < prices.get(Side.LAY)) {
                return Set.of(PREFIX + "loss");
            } else {
                throw new IllegalStateException();
            }
        } else {
            return Set.of(PREFIX + "none");
        }
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/CounterBetCategorizer.kt
package cz.fb.manaus.reactor.categorizer


import cz.fb.manaus.core.category.BetCoverage
import cz.fb.manaus.core.category.categorizer.RealizedBetCategorizer
import cz.fb.manaus.core.repository.domain.Price
import cz.fb.manaus.core.repository.domain.RealizedBet
import cz.fb.manaus.core.repository.domain.Side
import org.springframework.stereotype.Component

@Component
class CounterBetCategorizer : RealizedBetCategorizer {

    override val isSimulationSupported: Boolean = false

    override fun getCategories(realizedBet: RealizedBet, coverage: BetCoverage): Set<String> {
        val marketId = realizedBet.betAction.market.id
        val selectionId = realizedBet.selectionId
        val side = realizedBet.price.side
        val counterSide = side.opposite
        val bets = coverage.getBets(marketId, selectionId, counterSide)
        val avgCounter = bets
                .map { it.price }
                .map { it.price }
                .average()
        return if (avgCounter > 0) {
            val price = realizedBet.price.price
            val prices = mapOf(side to price, counterSide to avgCounter)
            when {
                Price.priceEq(avgCounter, price) -> setOf(PREFIX + "zero")
                prices[Side.BACK]!! > prices[Side.LAY]!! -> setOf(PREFIX + "profit")
                prices[Side.BACK]!! < prices[Side.LAY]!! -> setOf(PREFIX + "loss")
                else -> throw IllegalStateException()
            }
        } else {
            setOf(PREFIX + "none")
        }
    }

    companion object {
        const val PREFIX = "counter_"
    }

}
